File: android/app/src/main/java/com/utex/deadline/EventStore.kt
package com.utex.deadline

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import org.json.JSONArray
import org.json.JSONObject

object EventStore {
    private const val PREFS = "ute_deadline_secure_prefs"
    private const val LEGACY_PREFS = "ute_deadline_prefs"
    private const val KEY_ICAL_URL = "ical_url"
    private const val KEY_EVENTS = "events_json"
    private const val KEY_KNOWN_IDS = "known_ids"
    private const val KEY_LAST_SYNC = "last_sync"
    private const val KEY_DAILY_SUMMARY_TOUCHED = "daily_summary_touched"
    private const val KEY_DAILY_SUMMARY_ENABLED = "daily_summary_enabled"
    private const val KEY_DAILY_SUMMARY_HOUR = "daily_summary_hour"
    private const val KEY_DAILY_SUMMARY_MINUTE = "daily_summary_minute"
    private const val KEY_DAILY_SUMMARY_DAYS = "daily_summary_days"
    private const val KEY_DONE_IDS = "done_ids"
    private const val KEY_REMINDER_OFFSETS = "reminder_offsets"
    private const val KEY_PENDING_NOTIFICATIONS = "pending_notifications_json"
    const val ALL_DAYS_MASK = 0b1111111
    private val DEFAULT_REMINDER_MINUTES = listOf(24L * 60L, 12L * 60L, 60L)
    private val ALLOWED_REMINDER_MINUTES = listOf(2L * 24L * 60L, 24L * 60L, 12L * 60L, 3L * 60L, 60L, 30L)
    @Volatile
    private var cachedPrefs: SharedPreferences? = null
    @Volatile
    private var cachedEvents: List<DeadlineEvent>? = null
    @Volatile
    private var cachedDoneIds: Set<String>? = null

    fun getIcalUrl(context: Context): String {
        return prefs(context).getString(KEY_ICAL_URL, "").orEmpty()
    }

    fun setIcalUrl(context: Context, url: String) {
        prefs(context).edit().putString(KEY_ICAL_URL, url.trim()).apply()
    }

    fun clearConnection(context: Context) {
        prefs(context).edit()
            .remove(KEY_ICAL_URL)
            .remove(KEY_EVENTS)
            .remove(KEY_KNOWN_IDS)
            .remove(KEY_LAST_SYNC)
            .remove(KEY_DAILY_SUMMARY_TOUCHED)
            .remove(KEY_DONE_IDS)
            .remove(KEY_PENDING_NOTIFICATIONS)
            .putBoolean(KEY_DAILY_SUMMARY_ENABLED, false)
            .apply()
        cachedEvents = emptyList()
        cachedDoneIds = emptySet()
        clearLegacyConnection(context)
    }

    fun getLastSync(context: Context): Long {
        return prefs(context).getLong(KEY_LAST_SYNC, 0L)
    }

    fun setLastSync(context: Context, millis: Long) {
        prefs(context).edit().putLong(KEY_LAST_SYNC, millis).apply()
    }

    fun isDailySummaryEnabled(context: Context): Boolean {
        val preferences = prefs(context)
        return preferences.getString(KEY_ICAL_URL, "").orEmpty().isNotBlank() &&
            preferences.getBoolean(KEY_DAILY_SUMMARY_ENABLED, false)
    }

    fun setDailySummaryEnabled(context: Context, enabled: Boolean) {
        prefs(context).edit()
            .putBoolean(KEY_DAILY_SUMMARY_TOUCHED, true)
            .putBoolean(KEY_DAILY_SUMMARY_ENABLED, enabled)
            .apply()
    }

    fun prepareDailySummaryAfterSetup(context: Context) {
        val preferences = prefs(context)
        if (preferences.getBoolean(KEY_DAILY_SUMMARY_TOUCHED, false)) return
        preferences.edit()
            .putInt(KEY_DAILY_SUMMARY_HOUR, getDailySummaryHour(context))
            .putInt(KEY_DAILY_SUMMARY_MINUTE, getDailySummaryMinute(context))
            .putInt(KEY_DAILY_SUMMARY_DAYS, getDailySummaryDaysMask(context))
            .apply()
    }

    fun getDailySummaryHour(context: Context): Int {
        return prefs(context).getInt(KEY_DAILY_SUMMARY_HOUR, 6)
    }

    fun getDailySummaryMinute(context: Context): Int {
        return prefs(context).getInt(KEY_DAILY_SUMMARY_MINUTE, 0)
    }

    fun setDailySummaryTime(context: Context, hour: Int, minute: Int) {
        prefs(context).edit()
            .putBoolean(KEY_DAILY_SUMMARY_TOUCHED, true)
            .putBoolean(KEY_DAILY_SUMMARY_ENABLED, true)
            .putInt(KEY_DAILY_SUMMARY_HOUR, hour.coerceIn(0, 23))
            .putInt(KEY_DAILY_SUMMARY_MINUTE, minute.coerceIn(0, 59))
            .apply()
    }

    fun getDailySummaryDaysMask(context: Context): Int {
        return prefs(context).getInt(KEY_DAILY_SUMMARY_DAYS, ALL_DAYS_MASK).takeIf { it != 0 } ?: ALL_DAYS_MASK
    }

    fun setDailySummaryDaysMask(context: Context, daysMask: Int) {
        val safeMask = daysMask and ALL_DAYS_MASK
        prefs(context).edit()
            .putBoolean(KEY_DAILY_SUMMARY_TOUCHED, true)
            .putBoolean(KEY_DAILY_SUMMARY_ENABLED, safeMask != 0)
            .putInt(KEY_DAILY_SUMMARY_DAYS, safeMask)
            .apply()
    }

    fun isDailySummaryAllowedToday(context: Context): Boolean {
        val dayValue = java.time.LocalDate.now(java.time.ZoneId.of("Asia/Ho_Chi_Minh")).dayOfWeek.value
        val bit = 1 shl (dayValue - 1)
        return getDailySummaryDaysMask(context) and bit != 0
    }

    fun loadEvents(context: Context): List<DeadlineEvent> {
        cachedEvents?.let { return it }
        val json = prefs(context).getString(KEY_EVENTS, "[]").orEmpty()
        return try {
            val arr = JSONArray(json)
            (0 until arr.length()).mapNotNull { i ->
                val obj = arr.optJSONObject(i) ?: return@mapNotNull null
                DeadlineEvent(
                    id = obj.optString("id"),
                    title = obj.optString("title"),
                    startAtMillis = obj.optLong("startAtMillis"),
                    sourceUrl = obj.optString("sourceUrl").takeIf { it.isNotBlank() },
                    rawType = obj.optString("rawType").takeIf { it.isNotBlank() },
                    description = obj.optString("description").takeIf { it.isNotBlank() }
                )
            }.sortedBy { it.startAtMillis }.also {
                cachedEvents = it
            }
        } catch (_: Exception) {
            emptyList<DeadlineEvent>().also {
                cachedEvents = it
            }
        }
    }

    fun saveEvents(context: Context, events: List<DeadlineEvent>) {
        val arr = JSONArray()
        val sortedEvents = events.sortedBy { it.startAtMillis }
        sortedEvents.forEach { event ->
            arr.put(JSONObject().apply {
                put("id", event.id)
                put("title", event.title)
                put("startAtMillis", event.startAtMillis)
                put("sourceUrl", event.sourceUrl ?: "")
                put("rawType", event.rawType ?: "")
                put("description", event.description ?: "")
            })
        }
        prefs(context).edit().putString(KEY_EVENTS, arr.toString()).apply()
        cachedEvents = sortedEvents
        clearDoneForMissingEvents(context, sortedEvents.map { it.id }.toSet())
    }

    fun getKnownIds(context: Context): MutableSet<String> {
        return prefs(context).getStringSet(KEY_KNOWN_IDS, emptySet())?.toMutableSet() ?: mutableSetOf()
    }

    fun saveKnownIds(context: Context, ids: Set<String>) {
        prefs(context).edit().putStringSet(KEY_KNOWN_IDS, ids.toSet()).apply()
    }

    fun resetKnownIds(context: Context) {
        prefs(context).edit().remove(KEY_KNOWN_IDS).apply()
    }

    fun loadPendingDeadlineNotifications(context: Context): List<PendingDeadlineNotification> {
        val json = prefs(context).getString(KEY_PENDING_NOTIFICATIONS, "[]").orEmpty()
        return try {
            val arr = JSONArray(json)
            (0 until arr.length()).mapNotNull { i ->
                val obj = arr.optJSONObject(i) ?: return@mapNotNull null
                val type = obj.optString("type").takeIf { it == "new" || it == "changed" } ?: return@mapNotNull null
                val event = obj.optJSONObject("event")?.let { pendingEventFromJson(it) } ?: return@mapNotNull null
                val key = obj.optString("key").takeIf { it.isNotBlank() } ?: "$type-${event.id}"
                PendingDeadlineNotification(key, type, event)
            }
        } catch (_: Exception) {
            emptyList()
        }
    }

    fun upsertPendingDeadlineNotifications(context: Context, notifications: List<PendingDeadlineNotification>) {
        if (notifications.isEmpty()) return
        val merged = LinkedHashMap<String, PendingDeadlineNotification>()
        loadPendingDeadlineNotifications(context).forEach { merged[it.key] = it }
        notifications.forEach { merged[it.key] = it }
        savePendingDeadlineNotifications(context, merged.values.toList())
    }

    fun savePendingDeadlineNotifications(context: Context, notifications: List<PendingDeadlineNotification>) {
        val arr = JSONArray()
        notifications.distinctBy { it.key }.forEach { pending ->
            arr.put(JSONObject().apply {
                put("key", pending.key)
                put("type", pending.type)
                put("event", eventToJson(pending.event))
            })
        }
        prefs(context).edit().putString(KEY_PENDING_NOTIFICATIONS, arr.toString()).apply()
    }

    fun getDoneIds(context: Context): Set<String> {
        cachedDoneIds?.let { return it }
        return (prefs(context).getStringSet(KEY_DONE_IDS, emptySet())?.toSet() ?: emptySet()).also {
            cachedDoneIds = it
        }
    }

    fun isDone(context: Context, eventId: String): Boolean {
        return eventId in getDoneIds(context)
    }

    fun setDone(context: Context, eventId: String, done: Boolean) {
        val ids = getDoneIds(context).toMutableSet()
        if (done) {
            ids += eventId
        } else {
            ids -= eventId
        }
        val nextIds = ids.toSet()
        prefs(context).edit().putStringSet(KEY_DONE_IDS, nextIds).apply()
        cachedDoneIds = nextIds
    }

    fun getReminderOffsetOptions(): List<Long> = ALLOWED_REMINDER_MINUTES

    fun getReminderOffsetsMinutes(context: Context): List<Long> {
        val saved = prefs(context).getString(KEY_REMINDER_OFFSETS, null)
        val parsed = saved
            ?.split(',')
            ?.mapNotNull { it.trim().toLongOrNull() }
            ?.filter { it in ALLOWED_REMINDER_MINUTES }
            ?.distinct()
            .orEmpty()
        val selected = parsed.ifEmpty { DEFAULT_REMINDER_MINUTES }
        return ALLOWED_REMINDER_MINUTES.filter { it in selected }
    }

    fun isReminderOffsetEnabled(context: Context, minutes: Long): Boolean {
        return minutes in getReminderOffsetsMinutes(context)
    }

    fun setReminderOffsetEnabled(context: Context, minutes: Long, enabled: Boolean) {
        if (minutes !in ALLOWED_REMINDER_MINUTES) return
        val selected = getReminderOffsetsMinutes(context).toMutableSet()
        if (enabled) {
            selected += minutes
        } else {
            selected -= minutes
        }
        val safeSelected = selected.ifEmpty { setOf(60L) }
        val saved = ALLOWED_REMINDER_MINUTES
            .filter { it in safeSelected }
            .joinToString(",")
        prefs(context).edit().putString(KEY_REMINDER_OFFSETS, saved).apply()
    }

    fun reminderOptionLabel(minutes: Long): String {
        return when (minutes) {
            2L * 24L * 60L -> "2 ngày"
            24L * 60L -> "1 ngày"
            12L * 60L -> "12 giờ"
            3L * 60L -> "3 giờ"
            60L -> "1 giờ"
            30L -> "30 phút"
            else -> "$minutes phút"
        }
    }

    fun reminderLeadLabel(minutes: Long): String {
        return "${reminderOptionLabel(minutes)} trước hạn"
    }

    fun reminderOffsetsText(context: Context): String {
        return getReminderOffsetsMinutes(context)
            .joinToString(", ") { reminderOptionLabel(it) }
            .ifBlank { "1 giờ" }
    }

    private fun prefs(context: Context): SharedPreferences {
        cachedPrefs?.let { return it }
        return synchronized(this) {
            cachedPrefs ?: run {
                val appContext = context.applicationContext
                val securePrefs = securePrefs(appContext)
                val preferences = if (securePrefs != null) {
                    migrateLegacyPrefs(appContext, securePrefs)
                    securePrefs
                } else {
                    legacyPrefs(appContext)
                }
                cachedPrefs = preferences
                preferences
            }
        }
    }

    private fun securePrefs(context: Context): SharedPreferences? {
        return try {
            val masterKey = MasterKey.Builder(context)
                .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                .build()
            EncryptedSharedPreferences.create(
                context,
                PREFS,
                masterKey,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
            )
        } catch (_: Exception) {
            null
        }
    }

    private fun legacyPrefs(context: Context): SharedPreferences {
        return context.getSharedPreferences(LEGACY_PREFS, Context.MODE_PRIVATE)
    }

    private fun migrateLegacyPrefs(context: Context, securePrefs: SharedPreferences) {
        val legacyPrefs = legacyPrefs(context)
        if (!legacyPrefs.contains(KEY_ICAL_URL) || securePrefs.contains(KEY_ICAL_URL)) return

        val editor = securePrefs.edit()
        listOf(
            KEY_ICAL_URL,
            KEY_EVENTS,
            KEY_LAST_SYNC,
            KEY_DAILY_SUMMARY_ENABLED,
            KEY_DAILY_SUMMARY_TOUCHED,
            KEY_DAILY_SUMMARY_HOUR,
            KEY_DAILY_SUMMARY_MINUTE,
            KEY_DAILY_SUMMARY_DAYS,
            KEY_REMINDER_OFFSETS,
            KEY_PENDING_NOTIFICATIONS
        ).forEach { key ->
            when (val value = legacyPrefs.all[key]) {
                is String -> editor.putString(key, value)
                is Boolean -> editor.putBoolean(key, value)
                is Int -> editor.putInt(key, value)
                is Long -> editor.putLong(key, value)
            }
        }
        legacyPrefs.getStringSet(KEY_KNOWN_IDS, null)?.let {
            editor.putStringSet(KEY_KNOWN_IDS, it)
        }
        legacyPrefs.getStringSet(KEY_DONE_IDS, null)?.let {
            editor.putStringSet(KEY_DONE_IDS, it)
        }
        editor.apply()
        clearLegacyConnection(context)
    }

    private fun clearDoneForMissingEvents(context: Context, eventIds: Set<String>) {
        val doneIds = getDoneIds(context)
        val keptIds = doneIds.intersect(eventIds)
        if (keptIds.size != doneIds.size) {
            prefs(context).edit().putStringSet(KEY_DONE_IDS, keptIds).apply()
            cachedDoneIds = keptIds
        }
    }

    private fun eventToJson(event: DeadlineEvent): JSONObject {
        return JSONObject().apply {
            put("id", event.id)
            put("title", event.title)
            put("startAtMillis", event.startAtMillis)
            put("sourceUrl", event.sourceUrl ?: "")
            put("rawType", event.rawType ?: "")
            put("description", event.description ?: "")
        }
    }

    private fun pendingEventFromJson(obj: JSONObject): DeadlineEvent? {
        val id = obj.optString("id").takeIf { it.isNotBlank() } ?: return null
        val title = obj.optString("title").takeIf { it.isNotBlank() } ?: return null
        val startAtMillis = obj.optLong("startAtMillis", 0L)
        if (startAtMillis <= 0L) return null
        return DeadlineEvent(
            id = id,
            title = title,
            startAtMillis = startAtMillis,
            sourceUrl = obj.optString("sourceUrl").takeIf { it.isNotBlank() },
            rawType = obj.optString("rawType").takeIf { it.isNotBlank() },
            description = obj.optString("description").takeIf { it.isNotBlank() }
        )
    }

    private fun clearLegacyConnection(context: Context) {
        legacyPrefs(context).edit()
            .remove(KEY_ICAL_URL)
            .remove(KEY_EVENTS)
            .remove(KEY_KNOWN_IDS)
            .remove(KEY_LAST_SYNC)
            .remove(KEY_DONE_IDS)
            .remove(KEY_PENDING_NOTIFICATIONS)
            .apply()
    }
}

File: ios-local/UTEDeadlineLocal/Services/EventStore.swift
import Foundation

final class EventStore {
    static let shared = EventStore()

    private enum Keys {
        static let events = "events_json"
        static let knownIds = "known_ids"
        static let doneIds = "done_ids"
        static let pendingNotifications = "pending_notifications_json"
        static let lastSync = "last_sync"
        static let dailySummaryTouched = "daily_summary_touched"
        static let dailySummaryEnabled = "daily_summary_enabled"
        static let dailySummaryHour = "daily_summary_hour"
        static let dailySummaryMinute = "daily_summary_minute"
        static let dailySummaryDays = "daily_summary_days"
        static let reminderOffsets = "reminder_offsets"
        static let customReminderOffsets = "custom_reminder_offsets"
    }

    let presetReminderMinutes: [Int] = [
        7 * 24 * 60,
        3 * 24 * 60,
        2 * 24 * 60,
        24 * 60,
        12 * 60,
        6 * 60,
        3 * 60,
        60,
        30
    ]
    private let defaultReminderMinutes: [Int] = [24 * 60, 12 * 60, 60]
    private let allDaysMask = 0b1111111
    private let defaults = UserDefaults.standard
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    private init() {}

    var iCalURL: String {
        KeychainStore.getIcalURL()
    }

    func setIcalURL(_ url: String) throws {
        try KeychainStore.setIcalURL(url)
    }

    func clearConnection() {
        KeychainStore.clearIcalURL()
        [
            Keys.events,
            Keys.knownIds,
            Keys.doneIds,
            Keys.pendingNotifications,
            Keys.lastSync,
            Keys.dailySummaryTouched
        ].forEach { defaults.removeObject(forKey: $0) }
        defaults.set(false, forKey: Keys.dailySummaryEnabled)
    }

    func loadEvents() -> [DeadlineEvent] {
        decode([DeadlineEvent].self, key: Keys.events, fallback: [])
            .sorted { $0.startAtMillis < $1.startAtMillis }
    }

    func saveEvents(_ events: [DeadlineEvent]) {
        let sorted = events.sorted { $0.startAtMillis < $1.startAtMillis }
        encode(sorted, key: Keys.events)
        clearDoneForMissingEvents(Set(sorted.map(\.id)))
    }

    func knownIds() -> Set<String> {
        Set(defaults.stringArray(forKey: Keys.knownIds) ?? [])
    }

    func saveKnownIds(_ ids: Set<String>) {
        defaults.set(Array(ids), forKey: Keys.knownIds)
    }

    func doneIds() -> Set<String> {
        Set(defaults.stringArray(forKey: Keys.doneIds) ?? [])
    }

    func isDone(_ id: String) -> Bool {
        doneIds().contains(id)
    }

    func setDone(_ id: String, done: Bool) {
        var ids = doneIds()
        if done {
            ids.insert(id)
        } else {
            ids.remove(id)
        }
        defaults.set(Array(ids), forKey: Keys.doneIds)
    }

    var lastSync: Date? {
        let seconds = defaults.double(forKey: Keys.lastSync)
        return seconds > 0 ? Date(timeIntervalSince1970: seconds) : nil
    }

    func setLastSync(_ date: Date) {
        defaults.set(date.timeIntervalSince1970, forKey: Keys.lastSync)
    }

    func resetKnownIds() {
        defaults.removeObject(forKey: Keys.knownIds)
    }

    func loadPendingNotifications() -> [PendingDeadlineNotification] {
        decode([PendingDeadlineNotification].self, key: Keys.pendingNotifications, fallback: [])
    }

    func upsertPendingNotifications(_ notifications: [PendingDeadlineNotification]) {
        guard !notifications.isEmpty else { return }
        var merged: [String: PendingDeadlineNotification] = [:]
        loadPendingNotifications().forEach { merged[$0.key] = $0 }
        notifications.forEach { merged[$0.key] = $0 }
        savePendingNotifications(Array(merged.values))
    }

    func savePendingNotifications(_ notifications: [PendingDeadlineNotification]) {
        var seen = Set<String>()
        let unique = notifications.filter { seen.insert($0.key).inserted }
        encode(unique, key: Keys.pendingNotifications)
    }

    func prepareDailySummaryDefaultsAfterSetup() {
        guard !defaults.bool(forKey: Keys.dailySummaryTouched) else { return }
        defaults.register(defaults: [
            Keys.dailySummaryHour: 6,
            Keys.dailySummaryMinute: 0,
            Keys.dailySummaryDays: allDaysMask,
            Keys.dailySummaryEnabled: false
        ])
    }

    var isDailySummaryEnabled: Bool {
        get { !iCalURL.isEmpty && defaults.bool(forKey: Keys.dailySummaryEnabled) }
        set {
            defaults.set(true, forKey: Keys.dailySummaryTouched)
            defaults.set(newValue, forKey: Keys.dailySummaryEnabled)
        }
    }

    var dailySummaryHour: Int {
        get { defaults.object(forKey: Keys.dailySummaryHour) as? Int ?? 6 }
        set { defaults.set(max(0, min(23, newValue)), forKey: Keys.dailySummaryHour) }
    }

    var dailySummaryMinute: Int {
        get { defaults.object(forKey: Keys.dailySummaryMinute) as? Int ?? 0 }
        set { defaults.set(max(0, min(59, newValue)), forKey: Keys.dailySummaryMinute) }
    }

    var dailySummaryDaysMask: Int {
        get {
            let value = defaults.object(forKey: Keys.dailySummaryDays) as? Int ?? allDaysMask
            return value == 0 ? allDaysMask : value
        }
        set {
            defaults.set(true, forKey: Keys.dailySummaryTouched)
            let safe = newValue & allDaysMask
            defaults.set(safe, forKey: Keys.dailySummaryDays)
            defaults.set(safe != 0, forKey: Keys.dailySummaryEnabled)
        }
    }

    func isDailySummaryAllowedToday(date: Date = Date()) -> Bool {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "Asia/Ho_Chi_Minh") ?? .current
        let weekday = calendar.component(.weekday, from: date)
        let mondayBasedIndex = (weekday + 5) % 7
        return dailySummaryDaysMask & (1 << mondayBasedIndex) != 0
    }

    var allowedReminderMinutes: [Int] {
        allReminderOffsetOptions()
    }

    func allReminderOffsetOptions() -> [Int] {
        Array(Set(presetReminderMinutes + customReminderOffsets()))
            .filter { $0 > 0 }
            .sorted(by: >)
    }

    func customReminderOffsets() -> [Int] {
        (defaults.stringArray(forKey: Keys.customReminderOffsets) ?? [])
            .compactMap { Int($0) }
            .filter { $0 > 0 }
            .sorted(by: >)
    }

    func addCustomReminderOffset(_ minutes: Int) {
        guard minutes > 0 else { return }
        var custom = Set(customReminderOffsets())
        custom.insert(minutes)
        defaults.set(custom.sorted(by: >).map { String($0) }, forKey: Keys.customReminderOffsets)

        var selected = Set(reminderOffsetsMinutes())
        selected.insert(minutes)
        saveReminderOffsetsMinutes(Array(selected))
    }

    func removeCustomReminderOffset(_ minutes: Int) {
        var custom = Set(customReminderOffsets())
        custom.remove(minutes)
        defaults.set(custom.sorted(by: >).map { String($0) }, forKey: Keys.customReminderOffsets)

        if !presetReminderMinutes.contains(minutes) {
            var selected = Set(reminderOffsetsMinutes())
            selected.remove(minutes)
            if selected.isEmpty {
                selected.insert(60)
            }
            saveReminderOffsetsMinutes(Array(selected))
        }
    }

    func reminderOffsetsMinutes() -> [Int] {
        let allOptions = allReminderOffsetOptions()
        let saved = defaults.stringArray(forKey: Keys.reminderOffsets)?
            .compactMap { Int($0) }
            .filter { allOptions.contains($0) }
            ?? []
        let selected = saved.isEmpty ? defaultReminderMinutes : saved
        return allOptions.filter { selected.contains($0) }
    }

    func setReminderOffset(_ minutes: Int, enabled: Bool) {
        guard allReminderOffsetOptions().contains(minutes) else { return }
        var selected = Set(reminderOffsetsMinutes())
        if enabled {
            selected.insert(minutes)
        } else {
            selected.remove(minutes)
        }
        if selected.isEmpty {
            selected.insert(60)
        }
        saveReminderOffsetsMinutes(Array(selected))
    }

    func reminderOptionLabel(_ minutes: Int) -> String {
        switch minutes {
        case 7 * 24 * 60:
            return "7 ngày"
        case 3 * 24 * 60:
            return "3 ngày"
        case 2 * 24 * 60:
            return "2 ngày"
        case 24 * 60:
            return "1 ngày"
        case 12 * 60:
            return "12 giờ"
        case 6 * 60:
            return "6 giờ"
        case 3 * 60:
            return "3 giờ"
        case 60:
            return "1 giờ"
        case 30:
            return "30 phút"
        default:
            return customReminderLabel(minutes)
        }
    }

    private func saveReminderOffsetsMinutes(_ offsets: [Int]) {
        let allOptions = allReminderOffsetOptions()
        let safe = offsets
            .filter { $0 > 0 && allOptions.contains($0) }
            .sorted(by: >)
        defaults.set(safe.map { String($0) }, forKey: Keys.reminderOffsets)
    }

    private func customReminderLabel(_ minutes: Int) -> String {
        let days = minutes / (24 * 60)
        let hours = (minutes % (24 * 60)) / 60
        let mins = minutes % 60

        switch (days, hours, mins) {
        case let (days, 0, 0) where days > 0:
            return "\(days) ngày"
        case let (days, hours, 0) where days > 0:
            return "\(days) ngày \(hours) giờ"
        case let (days, hours, mins) where days > 0:
            return "\(days) ngày \(hours) giờ \(mins) phút"
        case let (0, hours, 0) where hours > 0:
            return "\(hours) giờ"
        case let (0, hours, mins) where hours > 0:
            return "\(hours) giờ \(mins) phút"
        default:
            return "\(mins) phút"
        }
    }

    private func clearDoneForMissingEvents(_ eventIds: Set<String>) {
        let kept = doneIds().intersection(eventIds)
        defaults.set(Array(kept), forKey: Keys.doneIds)
    }

    private func encode<T: Encodable>(_ value: T, key: String) {
        if let data = try? encoder.encode(value) {
            defaults.set(data, forKey: key)
        }
    }

    private func decode<T: Decodable>(_ type: T.Type, key: String, fallback: T) -> T {
        guard let data = defaults.data(forKey: key), let value = try? decoder.decode(type, from: data) else {
            return fallback
        }
        return value
    }
}
